packages/devextreme/js/__internal/core/state_manager/controller_registry.ts

```diff
@@ -0,0 +1,66 @@
+import { EventEmitter } from './event_emitter';
+import type { IController, IControllerRegistry, ILogger } from './interfaces';
+
+export class ControllerRegistry implements IControllerRegistry {
+  private controllers: Record<string, IController> = {};
+
+  private readonly logger: ILogger;
+
+  private readonly controllerRegisteredEmitter: EventEmitter<
+    (controller: IController, id: string) => void>;
+
+  constructor(logger: ILogger) {
+    this.logger = logger;
+    this.controllerRegisteredEmitter = new EventEmitter<(controller: IController,
+    id: string) => void
+      >(
+      'controllerRegistered',
+      logger,
+      );
+  }
+
+  registerController(controller: IController, controllerId: string): string {
+    if (!controller) {
+      this.logger.error('Controller cannot be null or undefined');
+
+      return '';
+    }
+
+    if (this.controllers[controllerId]) {
+      this.logger.debug(`Controller with ID ${controllerId} is already registered. Overwriting.`);
+    }
+
+    this.controllers[controllerId] = controller;
+    this.logger.debug(`Registered controller: ${controllerId}`);
+
+    this.controllerRegisteredEmitter.emit(controller, controllerId);
+
+    return controllerId;
+  }
+
+  getController(id: string): IController | null {
```

Comment:
I belive `undefined` is more conventional in TS in such cases
```suggestion
  getController(id: string): IController | undefined {
```

Reply:
I set it manually, so the `null` means a concise decision, not an accidentally obtained value.

`null` semantic:

> The **`null`** value represents the intentional absence of any object value.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/null

Reply:
Consider another analogy, which may be more relevant in this case, because we're speaking about a getter:

- [Array.find](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) returns `undefined` if no value found
- [Map.get](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get) returns `undefined` as well

Reply:
> the null means a concise decision, not an accidentally obtained value.

Since `undefined` is not a valid value for controller, there is no real difference for consuming code

Reply:
Good points, changed.